src/cargo/ops/registry.rs

```diff
@@ -374,6 +380,70 @@ fn transmit(
     Ok(())
 }
 
+fn wait_for_publish(
+    config: &Config,
+    registry_src: SourceId,
+    pkg: &Package,
+    timeout: std::time::Duration,
+) -> CargoResult<()> {
+    let version_req = format!("={}", pkg.version());
+    let mut source = registry_src.load(config, &HashSet::new())?;
+    let source_description = source.describe();
+    let query = Dependency::parse(pkg.name(), Some(&version_req), registry_src)?;
+
+    let now = std::time::Instant::now();
+    let sleep_time = std::time::Duration::from_secs(1);
+    let mut logged = false;
+    loop {
+        {
+            let _lock = config.acquire_package_cache_lock()?;
+            // Force re-fetching the source
+            //
+            // As pulling from a git source is expensive, we track when we've done it within the
+            // process to only do it once, but we are one of the rare cases that needs to do it
+            // multiple times
+            config.updated_sources().remove(&registry_src);
```

Comment:
This was a surprising find and might be relevant for #1169 depending on how its implemented.

cargo makes the assumption its doing things in one pass and little ever becomes invalidated but `publish` breaks one of those cases when using a git remote and I have to hack in this line to invalidate things.

This had been masked because originally no other tests worked.  A previous PR got sparse index tests working but still not git indexes.  I had chalked the timeouts to the git index tests not working but this was also broken